lms/djangoapps/shoppingcart/views.py

```diff
@@ -352,8 +351,15 @@ def register_code_redemption(request, registration_code):
         course = get_course_by_id(getattr(course_registration, 'course_id'), depth=0)
         if reg_code_is_valid and not reg_code_already_redeemed:
             #now redeem the reg code.
-            RegistrationCodeRedemption.create_invoice_generated_registration_redemption(course_registration, request.user)
-            CourseEnrollment.enroll(request.user, course.id)
+            redemption = RegistrationCodeRedemption.create_invoice_generated_registration_redemption(course_registration, request.user)
+
+            kwargs = {}
+            if course_registration.mode_slug is not None:
+                kwargs['mode'] = course_registration.mode_slug
+
+            redemption.course_enrollment = CourseEnrollment.enroll(request.user, course.id, **kwargs)
```

Comment:
I'll double check with Griff but I believe we wanted to adhere to expiration dates in general, so this should account for that. 

No worries if you see anything that isn't a part of this PR: anything that will make the code more robust is better.